tls/s2n_connection.c

```diff
@@ -1708,3 +1710,137 @@ int s2n_connection_get_key_update_counts(struct s2n_connection *conn,
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_INVALID_STATE);
+
+    uint32_t context_length = 0;
+    POSIX_GUARD(s2n_connection_get_traffic_context_length(conn, &context_length));
+    POSIX_ENSURE(buffer_length >= context_length, S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer output = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&output, &context_blob));
+
+    POSIX_GUARD(s2n_stuffer_write_uint64(&output, S2N_TRAFFIC_CONTEXT_V1));
+    POSIX_GUARD(s2n_stuffer_write_uint8(&output, conn->actual_protocol_version));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_uint16(&output, conn->max_outgoing_fragment_length));
+
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        uint8_t secret_size = 0;
+        POSIX_GUARD(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.client_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.server_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.resumption_master_secret, secret_size));
+    } else {
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls12.master_secret, S2N_TLS_SECRET_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.client_random, S2N_TLS_RANDOM_DATA_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.server_random, S2N_TLS_RANDOM_DATA_LEN));
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_set_traffic_context(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer input = { 0 };
+    POSIX_GUARD(s2n_stuffer_init_written(&input, &context_blob));
+
+    uint64_t context_version = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint64(&input, &context_version));
+    /* No other version is supported currently */
+    POSIX_ENSURE(context_version == S2N_TRAFFIC_CONTEXT_V1, S2N_ERR_INVALID_STATE);
+
+    POSIX_GUARD(s2n_stuffer_read_uint8(&input, &conn->actual_protocol_version));
+    conn->server_protocol_version = conn->actual_protocol_version;
+    conn->client_protocol_version = conn->actual_protocol_version;
```

Comment:
I don't think it matters, although I agree, it is a little weird. I need to set a value for these fields, in particular, server_protocol_version gets used to determine[ which TLS version is written in the record header](https://github.com/aws/s2n-tls/blob/30a490170d9e560abaf9a38de8851b846aef479b/tls/s2n_record_write.c#L176).  We end up sending a much lower TLS version in the record header if this field isn't set. However, I don't think it's worth it to store these fields in the traffic context since we _should_ be able to assume the protocol version at this point.